### PR TITLE
[fpp] Initial contribution

### PR DESCRIPTION
**Pull request by computergeek1507**
# Description

Created a binding to Controls [Falcon Player (FPP)](https://github.com/FalconChristmas/fpp) Devices. Uses MQTT to update status.

https://community.openhab.org/t/fpp-binding/157478

### CODE REVIEW

File: bundles/org.openhab.binding.mqtt.fpp/src/main/feature/feature.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<features name="org.openhab.binding.mqtt.espmilighthub-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/feature/feature.xml:2]: espmilighthub seems like a copy/paste error, change to fpp. Some more in this file.

**Review by lsiepel — CHANGES_REQUESTED** [review]
This PR is a WIP for some months, that is perfectly fine. There are some build failures and other issues. I did a quickscan and the most important one te mention is that the channel and type id's are not according to the naming convention (`lower-cvase-hyphen`)
For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention 

Hope you can proceed with this PR, let me know if you need anything.

File: bundles/org.openhab.binding.mqtt.fpp/src/main/feature/feature.xml (hunk 1 above)
Comment (computergeek1507) [reply bundles/org.openhab.binding.mqtt.fpp/src/main/feature/feature.xml]: fixed

File: bundles/org.openhab.binding.mqtt.fpp/pom.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.2.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.mqtt.fpp/pom.xml:10]:
```suggestion
    <version>4.3.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        return;
+    }
+
+    private @Nullable String executeGet(String url) {
+        String response = null;
+        try {
+            response = HttpUtil.executeUrl("GET", "http://" + config.playerIP + url, 5000);
+
+        } catch (Exception e) {
+            logger.error("Failed HTTP Post", e);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:252]: For reference: https://www.openhab.org/docs/developer/guidelines.html#f-logging
```suggestion
            logger.warn("Failed HTTP Post", e);
```

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        try {
+            Properties header = new Properties();
+            header.put("Accept", "application/json");
+            header.put("Content-Type", "application/json");
+            String response = HttpUtil.executeUrl("POST", "http://" + config.playerIP + url, header,
+                    new ByteArrayInputStream(json.getBytes()), "application/json", 5000);
+
+            return !response.isEmpty();
+        } catch (Exception e) {
+            logger.error("Failed HTTP Post", e);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:267]:
```suggestion
            logger.warn("Failed HTTP Post", e);
```

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        if (topic.endsWith(STATUS_TOPIC)) {
+            processIncomingState(state);
+        } else if (topic.endsWith(VERSION_TOPIC)) {
+            updateState(CHANNEL_VERSION, new StringType(state));
+            updateStatus(ThingStatus.ONLINE);
+        }
+    }
+
+    public ThingStatusInfo getBridgeStatus() {
+        Bridge b = getBridge();
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:201]: Use meaningfull names.
```suggestion
        Bridge bridge = getBridge();
```

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                    "Bridge is missing or offline, you need to setup a working MQTT broker first.");
+            return;
+        }
+        ThingHandler handler = localBridge.getHandler();
+        if (handler instanceof AbstractBrokerHandler abh) {
+            final MqttBrokerConnection connection;
+            try {
+                connection = abh.getConnectionAsync().get(500, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException | ExecutionException | TimeoutException ignored) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_UNINITIALIZED,
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:233]: Calling this method is initializing the connection (if not already) and returning it. The init envolves a tcp/udp connection to the broker inclukding possible auhtentication.
Long story short, i hope all systems respond in 500ms but i think this is a bit to tight. Maybe 5000 is better.

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+import tech.units.indriya.unit.Units;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:56]: Don't use this package.

Comment (lsiepel) [reply bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/handler/FPPPlayerHandler.java:56]:
```suggestion
import org.openhab.core.library.unit.Units;
```

File: bundles/org.openhab.binding.mqtt.fpp/README.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+| `bridging-enabled` | Switch | Is Recieving Bridge Data. |
+| `multisync-enabled` | Switch | Multisync Mode Enabled. |
+| `scheduler-current-playlist` | String (read only) | Scheduler Current Playlist. |
+| `scheduler-current-playlist-start` | String (read only) | Scheduler Current Playlist Start Time. |
+| `scheduler-current-playlist-end` | String (read only) | Scheduler Current Playlist End Time. |
+| `scheduler-current-playlist-stop-type` | String (read only) | Scheduler Current Playlist End Type. |
+| `scheduler-next-playlist` | String (read only) | Next Scheduled Playlist. |
+| `scheduler-next-playlist-start` | String (read only) | Next Scheduled Start Time. |
+
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.fpp/README.md:39]:
```suggestion
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks for this contribution. Basics of this binding already seem good, there are some comments and questions. Besides this:
- check compile warnings, there should be no warnings.
- check SAT, only the missing bridge with id broker is ok, the rest should be fixed. 

Let me know if you need anything to proceed, ping me when you are ready for another review round.

File: bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/FPPBindingConstants.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    public static final String CHANNEL_SCHEDULERCURRENTPLAYLISTEND = "scheduler-current-playlist-end";
+    public static final String CHANNEL_SCHEDULERCURRENTPLAYLISTSTOPTYPE = "scheduler-current-playlist-stop-type";
+    public static final String CHANNEL_SCHEDULERNEXTPLAYLIST = "scheduler-next-playlist";
+    public static final String CHANNEL_SCHEDULERNEXTPLAYLISTSTART = "scheduler-next-playlist-start";
+
+    // Status
+    public static final String CONNECTED = "connected";
+    public static final String CHANNEL_STATUS_NAME = "status_name";
+
+    public static final String TESTING = "testing";
```
[the remark this answers is not in the document]
Comment (computergeek1507) [reply bundles/org.openhab.binding.mqtt.fpp/src/main/java/org/openhab/binding/mqtt/fpp/internal/FPPBindingConstants.java:71]: this is the static string value for when FPP status is in Testing mode

**Comment by lsiepel** [comment]
There is one comment left open, besides that this binding should also be added to the [footer.xml here](https://github.com/openhab/openhab-addons/blob/cae65393d41b4abae692d79c452883f06663ccbd/features/openhab-addons/src/main/resources/footer.xml#L31)
Otherwise we are ready to merge.

**Comment by lsiepel** [comment]
@computergeek1507  We can merge this and get it in the next milestone if you add the last two bits: codewoner and footer.xml

**Review by lsiepel — APPROVED** [review]
Thanks! LGTM.

Now, you could add your binding's logo to the openHAB website. See https://next.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

**Pull request merged**